spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/EmbeddedHeadersMessageConverter.java

```diff
@@ -93,38 +95,52 @@ public static String decodeExceptionMessage(Message<?> requestMessage) {
 	 * have been promoted back to actual headers. The new payload is now the original
 	 * payload.
 	 *
-	 * @param message the message to extract headers
-	 * @param copyRequestHeaders boolean value to specify if original headers should be
-	 * copied
+	 * @param message            the message to extract headers
+	 * @param copyRequestHeaders boolean value to specify if the request headers should be
+	 *                           copied
 	 */
 	public MessageValues extractHeaders(Message<byte[]> message, boolean copyRequestHeaders) throws Exception {
-		byte[] bytes = message.getPayload();
-		ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+		return extractHeaders(message.getPayload(), copyRequestHeaders, message.getHeaders());
+	}
+
+	/**
+	 * Return a message where headers, that were originally embedded into the payload,
+	 * have been promoted back to actual headers. The new payload is now the original
+	 * payload.
+	 *
+	 * @param payload            the message payload
+	 * @param copyRequestHeaders boolean value to specify if the request headers should be copied
+	 * @param requestHeaders     the request headers to copy
+	 * @return the message with extracted headers
+	 * @throws Exception
+	 */
+	public MessageValues extractHeaders(byte[] payload, boolean copyRequestHeaders, MessageHeaders requestHeaders) throws Exception {
```

Comment:
We only need two public methods:

```
public MessageValues extractHeaders(Message<byte[]> message, boolean copyRequestHeaders) throws Exception
```
and
```
public MessageValues extractHeaders(byte[] payload) throws Exception
```

This method can be internal - in the future, this will need to be refactored as we don't need to pass both the boolean argument and the requestHeaders. Passing an empty or null set of request headers should be enough to signal that there's nothing to copy.